pilot/cmd/pilot-agent/status/server.go

```diff
@@ -756,6 +751,31 @@ func (s *Server) handleAppProbeHTTPGet(w http.ResponseWriter, req *http.Request,
 	w.WriteHeader(response.StatusCode)
 }
 
+// Host can either come from user explicitly setting the host in `httpHeaders`, or implicit defined to be IP:port.
+//
+// `httpHeaders`, in an edge case, can set Host multiple times due to header casing.
+// On older Kubernetes versions, this results in 2 Host headers, which Go (pilot-agent) rejects.
+// On newer versions (https://github.com/kubernetes/kubernetes/pull/117323), the first one is picked.
+//
+// Using the implicit hostname also has issues, though. The port will be seen as 15020 by Kubernetes. To handle this, we
+// need to overwrite it
+func determineHost(req *http.Request, prober *Prober) string {
+	for _, header := range prober.HTTPGet.HTTPHeaders {
+		if strings.EqualFold(header.Name, "Host") {
+			// Case 1: explicit set host.
+			// Case insensitive match with the first match winning
+			return header.Value
+		}
+	}
+	if host, _, err := net.SplitHostPort(req.Host); err == nil {
+		// Injected Prober already has port rewritten to a number, so IntValue is always correct
+		port := strconv.Itoa(prober.HTTPGet.Port.IntValue())
+		return net.JoinHostPort(host, port)
```

Comment:
This is also not right with port stripped if Host is "example.com:8080"

Reply:
How can this happen without hitting L767?

Note if you put `host: example.com` in the Pod YAML it will literally do `curl example.com` - not send to the pod at all.

Reply:
I mean Header[Host] = "example.com:8080"

Reply:
In this case `Host` will be removed from Headers in incoming request

Reply:
it's not reading front eh request but rather the original probe definition

Reply:
Actually not sure, there may still be some issue - need to try a few things

Reply:
Ok I did some more investigation. I think https://github.com/istio/istio/issues/46215 is broken everywhere (which is totally fine - not concerned), and that our PRs behavior identically in every case except

```
          httpGet:
            httpHeaders:
            - name: Host
              value: my-host:15020
```

Both take two different approaches though.

Given how many bugs we have had its good we have two implementations so we can validate each one and make sure we fully understand the problem.

So I want to verify:
* Do you agree both are correct? If not, one of us has a different understanding of the problem
* Do you want to go forward with yours or mine? I am fine either way

Also cc @ramaraochavali @jaellio to get some extra eyes -  I really want to make sure we get it right this time given this is attempt #8 or so :slightly_smiling_face:

Reply:
I was OOF last week - I'll take another look at this now.